File: backend/src/compose/compose.service.ts
import { Injectable } from '@nestjs/common';
import { InjectEntityManager, InjectRepository } from '@nestjs/typeorm';
import { EntityManager, Repository } from 'typeorm';
import { v4 } from 'uuid';
import { ComposeEntity } from './compose.entity';
import { ComposeVersionEntity } from './composeVersion.entity';

function serviceNameFromCompose(name: string) {
  return name.replace(/[^a-z0-9_-]/gi, '-').replace(/-+/g, '-');
}

@Injectable()
export class ComposeService {
  constructor(
    @InjectEntityManager()
    private readonly manager: EntityManager,
    @InjectRepository(ComposeEntity)
    private readonly composeRepo: Repository<ComposeEntity>,
  ) {}

  public async create(args: {
    name: string;
    directory: string;
    serviceEnabled: boolean;
    compose: string;
  }) {
    const compose = new ComposeEntity();
    compose.id = v4();
    compose.name = args.name;

    const composeVersion = new ComposeVersionEntity();
    composeVersion.id = v4();
    composeVersion.compose = Promise.resolve(compose);
    composeVersion.content = args.compose;
    composeVersion.directory = args.directory;
    composeVersion.serviceEnabled = args.serviceEnabled;
    composeVersion.serviceName = serviceNameFromCompose(args.name);

    await this.manager.transaction(async trx => {
      await trx.save(compose);
      await trx.save(composeVersion);
      compose.current = Promise.resolve(composeVersion);
      await trx.save(compose);
    });

    return compose.id;
  }

  public async update(
    id: string,
    args: {
      name: string;
      directory: string;
      serviceEnabled: boolean;
      compose: string;
    },
  ) {
    await this.manager.transaction(async trx => {
      const compose = await this.composeRepo.findOne(id);

      if (!compose) {
        throw new Error(`Compose with id ${id} not found`);
      }

      compose.name = args.name;

      const composeVersion = new ComposeVersionEntity();
      composeVersion.id = v4();
      composeVersion.compose = Promise.resolve(compose);
      composeVersion.content = args.compose;
      composeVersion.directory = args.directory;
      composeVersion.serviceEnabled = args.serviceEnabled;
      composeVersion.serviceName = serviceNameFromCompose(args.name);

      await trx.save(composeVersion);
      compose.current = Promise.resolve(composeVersion);
      await trx.save(compose);
    });
  }

  public async delete(id: string) {
    await this.composeRepo.update({ id: id }, { pendingDelete: true });
  }
}

File: backend2/Services/ComposeService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using ProtoBuf;
using supercompose;
using Z.EntityFramework.Plus;

namespace backend2.Services
{
  public class ComposeService
  {
    private readonly SupercomposeContext ctx;
    private readonly NodeUpdaterService nodeUpdater;

    public ComposeService(SupercomposeContext ctx, NodeUpdaterService nodeUpdater)
    {
      this.ctx = ctx;
      this.nodeUpdater = nodeUpdater;
    }

    public async Task<Guid> Create(string name, string directory, bool serviceEnabled, string content)
    {
      // The complexity in this method is needed because deferred CurrentId
      await using var transaction = await ctx.Database.BeginTransactionAsync();

      var composeId = Guid.NewGuid();
      var versionId = Guid.NewGuid();

      await ctx.Composes.AddAsync(new Compose
      {
        Id = composeId,
        Name = name,
        PendingDelete = false,
        CurrentId = versionId
      });

      await ctx.SaveChangesAsync();

      await ctx.ComposeVersions.AddAsync(new ComposeVersion
      {
        Id = versionId,
        Content = content,
        Directory = directory,
        ServiceEnabled = serviceEnabled,
        ServiceName = ServiceNameFromCompose(name),
        ComposeId = composeId
      });

      await ctx.SaveChangesAsync();
      await transaction.CommitAsync();

      return composeId;
    }

    public async Task Update(Guid id, string? name, string? directory, bool? serviceEnabled, string? content)
    {
      var compose = await ctx.Composes
        .Include(x => x.Current)
        .Include(x => x.Deployments)
        .FirstOrDefaultAsync(x => x.Id == id);

      if (compose == null) throw new ComposeNotFoundException();

      if (name != null) compose.Name = name;

      if (directory != null || serviceEnabled != null)
      {
        var version = new ComposeVersion
        {
          Id = Guid.NewGuid(),
          Content = content ?? compose.Current.Content,
          Directory = directory ?? compose.Current.Directory,
          ServiceEnabled = serviceEnabled ?? compose.Current.ServiceEnabled,
          ServiceName = name != null ? ServiceNameFromCompose(name) : compose.Current.ServiceName,
          ComposeId = compose.Id
        };

        await ctx.ComposeVersions.AddAsync(version);

        compose.CurrentId = version.Id;
      }

      await ctx.SaveChangesAsync();

      foreach (var deployment in compose.Deployments) await nodeUpdater.NotifyAboutNodeChange(deployment.NodeId.Value);
    }

    public async Task Delete(Guid id)
    {
      var compose = await ctx.Composes.Include(x => x.Deployments).FirstOrDefaultAsync(x => x.Id == id);

      if (compose == null) return;

      var nodeIds = compose.Deployments.Select(x => x.NodeId);

      ctx.Composes.Remove(compose);

      await ctx.SaveChangesAsync();

      foreach (var nodeId in nodeIds) await nodeUpdater.NotifyAboutNodeChange(nodeId.Value);
    }

    private string ServiceNameFromCompose(string name)
    {
      name = new Regex("[^a-z0-9_-]", RegexOptions.IgnoreCase).Replace(name, "-");
      name = new Regex("-+", RegexOptions.IgnoreCase).Replace(name, "-");
      return name;
    }
  }
}